src/IssueViz/Editor/LocationTagging/LocationTagger.cs

```diff
@@ -0,0 +1,188 @@
+    // * adds location tags for primary and secondary locations for a file
+    // * tracks buffer edits, re-calculate spans, and notifies the issues location store about the changes
+    internal class LocationTagger : ITagger<IIssueLocationTag>, IDisposable
+    {
+        private readonly ITextBuffer buffer;
+        private readonly IIssueLocationStore locationService;
+        private readonly IIssueSpanCalculator spanCalculator;
+        private readonly ILogger logger;
+
+        internal /* for testing */ IList<ITagSpan<IIssueLocationTag>> TagSpans { get; private set; }
+
+        public LocationTagger(ITextBuffer buffer, IIssueLocationStore locationService, IIssueSpanCalculator spanCalculator, ILogger logger)
+        {
+            this.buffer = buffer;
+            this.locationService = locationService;
+            this.spanCalculator = spanCalculator;
+            this.logger = logger;
+
+            FilePath = GetFileName(buffer);
+            UpdateTags();
+
+            locationService.IssuesChanged += OnIssuesChanged;
+            buffer.ChangedLowPriority += SafeOnBufferChange;
+        }
+
+        private void OnIssuesChanged(object sender, IssuesChangedEventArgs e)
+        {
+            if (!e.AnalyzedFiles.Contains(FilePath, StringComparer.OrdinalIgnoreCase))
+            {
+                return; // no changes in this file
+            }
+
+            UpdateTags();
+        }
+
+        private void UpdateTags()
+        {
+            var textSnapshot = buffer.CurrentSnapshot;
+
+            // Get the new locations and calculate the spans if necessary
+            var locations = locationService.GetLocations(FilePath);
+            EnsureSpansExist(locations, textSnapshot);
+
+            TagSpans = locations.Select(CreateTagSpan).ToList();
+
+            // Notify the editor that our set of tags has changed
+            var wholeSpan = new SnapshotSpan(textSnapshot, 0, textSnapshot.Length);
+            TagsChanged?.Invoke(this, new SnapshotSpanEventArgs(wholeSpan));
+        }
+
+        private void EnsureSpansExist(IEnumerable<IAnalysisIssueLocationVisualization> locVizs, ITextSnapshot currentSnapshot)
+        {
+            // All spans for issues in the primary file should have been calculated already (whether primary or secondary).
+            // However, there could be secondary locations relating to issues in other files that have not been calculated.
+            foreach (var locViz in locVizs)
+            {
+                if (!locViz.Span.HasValue)
+                {
+                    locViz.Span = spanCalculator.CalculateSpan(locViz.Location, currentSnapshot);
+                }
+            }
+        }
+
+        private static ITagSpan<IIssueLocationTag> CreateTagSpan(IAnalysisIssueLocationVisualization locViz) =>
+            new TagSpan<IIssueLocationTag>(locViz.Span.Value, new IssueLocationTag(locViz));
+
+        private string GetFileName(ITextBuffer textBuffer)
```

Comment:
How do you think you will handle file renames?
We used to subscribe to `Document.FileActionOccurred`, so it might be easier to just receive ITextDocument in the ctor instead of ITextBuffer, and then we can use it to get `Document.FilePath`.

Reply:
There's a separate ticket for that; I'll worry about it then.